src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensCache.cs

```diff
@@ -0,0 +1,130 @@
+            NextResultId = new Dictionary<Uri, int>();
+        }
+
+        /// <summary>
+        /// Updates the given document's token set cache. Removes old cache results if the document's
+        /// cache is full.
+        /// </summary>
+        public async Task UpdateCacheAsync(
+            Uri uri,
+            LSP.SemanticTokens tokens,
+            CancellationToken cancellationToken)
+        {
+            using (await _semaphore.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+            {
+                // Case 1: Document does not currently have any token sets cached. Create a cache
+                // for the document and return.
+                if (!Tokens.TryGetValue(uri, out var tokenSets))
+                {
+                    Tokens.Add(uri, new List<LSP.SemanticTokens> { tokens });
+                    return;
+                }
+
+                // Case 2: Document already has the maximum number of token sets cached. Remove the
+                // oldest token set from the cache, and then add the new token set (see case 3).
+                if (tokenSets.Count >= _maxCachesPerDoc)
+                {
+                    tokenSets.RemoveAt(0);
+                }
+
+                // Case 3: Document has less than the maximum number of token sets cached.
+                // Add new token set to cache.
+                tokenSets.Add(tokens);
+            }
+        }
+
+        /// <summary>
+        /// Returns the cached tokens for a given document URI and resultId.
+        /// Returns null if no match is found.
+        /// </summary>
+        public async Task<LSP.SemanticTokens?> GetCachedTokensAsync(
+            Uri uri,
+            string? resultId,
+            CancellationToken cancellationToken)
+        {
+            if (resultId == null)
+            {
+                return null;
+            }
+
+            using (await _semaphore.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+            {
+                if (!Tokens.TryGetValue(uri, out var tokenSets))
+                {
+                    return null;
+                }
+
+                // Return a non-null value only if the document's cache contains a token set with the resultId
+                // that the user is searching for.
+                return tokenSets.FirstOrDefault(t => t.ResultId != null && t.ResultId == resultId);
+            }
+        }
+
+        /// <summary>
+        /// Returns the next available resultId for a given document URI.
+        /// </summary>
+        public async Task<string> GetNextResultIdAsync(Uri uri, CancellationToken cancellationToken)
+        {
+            using (await _semaphore.DisposableWaitAsync(cancellationToken).ConfigureAwait(false))
+            {
+                NextResultId.TryGetValue(uri, out var resultId);
+                NextResultId[uri] = resultId + 1;
+                return resultId.ToString();
+            }
```

Comment:
we coudl make this much simpler just by having an int/long, and doing an Interlocked.Increment on it.  bam. easy, unique ids.